# Part 1: Excavating Dinosaurs 🦕

![Cover_Image_1](Cover_Image_1.jpg)

_“The Nation’s T. rex," a nearly complete skeleton of a Tyrannosaurus rex, is depicted devouring a Triceratops at the Smithsonian’s National Museum of Natural History. Cover Image from_ [Smithsonian Institution](https://www.si.edu/sites/default/files/newsdesk/press_releases/nhb2017-00021.jpg)

## Key Findings

- The dataset contains 1,042 distinct dinosaurs, highlighting an extensive biodiversity across the Mesozoic era, which reflects a variety of ecological niches and evolutionary adaptations.

- Supersaurus and Argentinosaurus, with lengths up to 35 meters, were identified as the largest dinosaurs, emphasizing the exceptional size achieved by sauropods during the Mesozoic.

- Ornithopods are the most frequently occurring dinosaurs in the dataset with 811 records, suggesting that their herbivorous lifestyle and potential social behaviors contributed significantly to their success and adaptability.

- Alberta and Wyoming are highlighted as regions with the highest frequency of dinosaur fossil discoveries, indicating their rich paleontological significance and the excellent preservation conditions of these areas.

- There is a statistically significant, yet weak, correlation between dinosaur size and age, suggesting that factors other than age, such as environmental conditions and ecological pressures, might have had a more substantial influence on the evolution of dinosaur size.

- There is a considerable amount of missing data, particularly in the categories of diet, type, and length, which poses challenges to drawing comprehensive conclusions about dinosaur ecology and physiology.

## 1.1. Overview

I am applying for a summer internship at a national museum for natural history. The museum recently created a database containing all dinosaur records of past field campaigns. My job is to dive into the fossil records to find some interesting insights, and advise the museum on the quality of the data.

**The objectives of this analysis includes:**

- Assess the diversity of the dataset, specifically quantifying the number of different dinosaur names recorded.
- Identify the largest dinosaur in the dataset and address any missing data that might impact the accuracy of findings.
- Determine which type of dinosaur appears most frequently in the dataset.
- Examine the size trend to assess if dinosaurs generally increased in size throughout their evolutionary history.
- Create an interactive map to visually represent each fossil record, enhancing data presentation and accessibility.
- Conduct a thorough investigation of the dataset to identify and present any other significant findings that emerge during the analysis.

## 1.2. Data Description

The dataset was collected from the [Paleobiology Database](https://paleobiodb.org/#/) website. The following table gives some descriptions of our column names or key variables:

| Column name | Description |
|---|---|
| occurence_no | The original occurrence number from the Paleobiology Database. |
| name | The accepted name of the dinosaur (usually the genus name, or the name of the footprint/egg fossil). |
| diet | The main diet (omnivorous, carnivorous, herbivorous). |
| type | The dinosaur type (small theropod, large theropod, sauropod, ornithopod, ceratopsian, armored dinosaur). |
| length_m | The maximum length, from head to tail, in meters. |
| max_ma | The age in which the first fossil records of the dinosaur were found, in million years. |
| min_ma | The age in which the last fossil records of the dinosaur were found, in million years. |
| region | The current region where the fossil record was found. |
| lng | The longitude where the fossil record was found. |
| lat | The latitude where the fossil record was found. |
| class | The taxonomical class of the dinosaur (Saurischia or Ornithischia). |
| family | The taxonomical family of the dinosaur (if known). |

The dataset was enriched with data from Wikipedia.

## 1.3. Initial Data Exploration

First, we import the essential Python packages needed for exploring and analyzing our dataset. Then, we conduct an initial exploration to identify potential issues, such as missing values and other inconsistencies with our dataset. This preliminary step helps us become acquainted with our data and enhances our understanding of its structure and content.

In [1]:
# Import necessary packages
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from scipy.stats import linregress

In [2]:
# Load the data into a Pandas DataFrame:
df = pd.read_csv('data/dinosaurs.csv')

In [3]:
# Preview the dataframe
df.head()

occurrence_no               name  ...       class             family
0         130294  Protarchaeopteryx  ...  Saurischia  Archaeopterygidae
1         130295        Caudipteryx  ...  Saurischia    Caudipterygidae
2         139242        Gorgosaurus  ...  Saurischia    Tyrannosauridae
3         139250        Gorgosaurus  ...  Saurischia    Tyrannosauridae
4         139252        Gorgosaurus  ...  Saurischia    Tyrannosauridae

[5 rows x 12 columns]

In [4]:
# Print information about the data, including the data types and number of non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4951 entries, 0 to 4950
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   occurrence_no  4951 non-null   int64  
 1   name           4951 non-null   object 
 2   diet           3596 non-null   object 
 3   type           3596 non-null   object 
 4   length_m       3568 non-null   float64
 5   max_ma         4951 non-null   float64
 6   min_ma         4951 non-null   float64
 7   region         4909 non-null   object 
 8   lng            4951 non-null   float64
 9   lat            4951 non-null   float64
 10  class          4951 non-null   object 
 11  family         3494 non-null   object 
dtypes: float64(5), int64(1), object(6)
memory usage: 464.3+ KB


In [5]:
# Check for missing values in each column
df.isnull().sum()

occurrence_no       0
name                0
diet             1355
type             1355
length_m         1383
max_ma              0
min_ma              0
region             42
lng                 0
lat                 0
class               0
family           1457
dtype: int64

In [6]:
# Show the summary statistics of our numerical variables
df.describe()

occurrence_no     length_m  ...          lng          lat
count   4.951000e+03  3568.000000  ...  4951.000000  4951.000000
mean    6.838323e+05     8.212688  ...   -37.048675    34.591448
std     3.021524e+05     6.629887  ...    84.591106    23.961138
min     1.302090e+05     0.450000  ...  -153.247498   -84.333336
25%     4.871225e+05     3.000000  ...  -108.258705    36.274439
50%     5.618050e+05     6.700000  ...   -96.099998    42.611198
75%     8.675665e+05    10.000000  ...    27.383331    47.745138
max     1.365954e+06    35.000000  ...   565.000000    78.101875

[8 rows x 6 columns]

In our initial exploration of our dinosaur dataset, we previewed 4951 entries across 12 columns, containing various attributes like names, diets, types, lengths, and geological ages. We identified missing values in several columns, notably 'diet', 'type', 'length_m', 'region', and 'family'. Our columns include various data types: integers, floats, and objects or strings, with numerical columns showing a wide range of values, such as 'length_m' spanning from 0.45 to 35 meters. This preliminary step provides a great start to our data exploration, which is key for guiding the next steps in our analysis.

## 1.4. Insights and Data Visualization

### 1. How many different dinosaur names are present in the data?

In [7]:
# Calculate the number of unique dinosaur names
unique_dinosaur_names = df['name'].nunique()

# Output the number of unique dinosaur names
print("Number of unique dinosaur names:", unique_dinosaur_names)

Number of unique dinosaur names: 1042


The presence of **1,042** unique dinosaur names in the dataset is indicative of the rich biodiversity that existed during the [Mesozoic era](https://www.britannica.com/science/Mesozoic-Era). This diversity in dinosaur genera reflects a wide array of ecological niches that these creatures occupied, ranging from different feeding behaviors (herbivorous, carnivorous, omnivorous) to varied habitats spread across the continents. Each name represents not just a distinct species but potentially different adaptations and evolutionary histories. The enormous number of unique dinosaur names can also highlight the effectiveness and extensive reach of paleontological explorations and fossil discoveries. It provides a vivid picture of the past biodiversity and allows us to understand more about the evolutionary processes that shaped the Mesozoic ecosystem.

### 2. Which was the largest dinosaur? What about missing data in the dataset?

In [8]:
# Identify all dinosaurs that have the maximum length recorded in the dataset
max_length = df['length_m'].max()
largest_dinosaurs = df[df['length_m'] == max_length][['name', 'length_m']]

# Count the number of missing values in the 'length_m' column
missing_length_data = df['length_m'].isna().sum()

# Display the results
print("Largest Dinosaurs:")
print(largest_dinosaurs)
print("Missing data in 'length_m' column:", missing_length_data)

Largest Dinosaurs:
                 name  length_m
570       Supersaurus      35.0
924       Supersaurus      35.0
1165  Argentinosaurus      35.0
1423      Supersaurus      35.0
1972      Supersaurus      35.0
Missing data in 'length_m' column: 1383


![Supersaurus](Supersaurus.jpg "Supersaurus")

_Image of the Supersaurus from_ [Wikimedia Commons](https://upload.wikimedia.org/wikipedia/commons/a/a4/Supersaurus_dinosaur.png)

**Supersaurus** and **Argentinosaurus** are the largest dinosaurs recorded, with estimated lengths of 35 meters. These findings align with our understanding that some of the largest dinosaurs belonged to the group of sauropods, which are known for their enormous body sizes, long necks, and tails. Both [Supersaurus](https://en.wikipedia.org/wiki/Supersaurus), predominantly found in North America, and [Argentinosaurus](https://en.wikipedia.org/wiki/Argentinosaurus), discovered in South America, exemplify the gigantism that sauropods achieved. The multiple entries for Supersaurus suggest that several fossils of comparable size have been discovered, which reinforces the accuracy of our measurements and the consistency of enormous size within this species. The presence of Argentinosaurus in this list corroborates its position among the largest dinosaurs known from skeletal reconstructions and fossil evidence.

![Argentinosaurus](Argentinosaurus.jpg "Argentinosaurus")

_Image of the Argentinosaurus from_ [Wikimedia Commons](https://upload.wikimedia.org/wikipedia/commons/e/e1/Argentinosaurus_BW.jpg)

The considerable number of **missing data entries (1383)** for `length_m` is a crucial aspect to consider. This highlights the inherent challenges in paleontology, such as incomplete fossil records and the difficulty in estimating the sizes of many dinosaurs. These size estimates can be subject to revision with new discoveries and more complete fossils. The missing data also underscores the potential for even larger or similarly large dinosaurs to exist that are yet to be documented or accurately measured in the dataset.

### 3. What dinosaur type has the most occurrences in this dataset?

In [9]:
# Count the occurrences of each dinosaur type
type_counts = df['type'].value_counts()

# Create a bar chart using Plotly
fig = px.bar(type_counts, 
             title="Which dinosaur type has the most occurrences?",
             labels={'value': 'Number of Occurrences', 'index': 'Dinosaur Type'},
             color_discrete_sequence=['black'])

fig.update_layout(
    plot_bgcolor='white',
    paper_bgcolor='white',
    showlegend=False
)

fig.show()

![Hadrosaurus_foulkii](Hadrosaurus_foulkii.jpg "Hadrosaurus foulkii")

_Image of Hadrosaurus foulkii from_ [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Hadrosaurus_foulkii_restoration.png)

[Ornithopods](https://www.geol.umd.edu/~tholtz/G104/lectures/104ornithop.html), known for their bipedal or quadrupedal herbivorous lifestyle, including well-known genera like Iguanodon and Hadrosaurus, show the highest number at 811 occurrences. This suggests a successful adaptation strategy, possibly due to their varied diet and social behaviors which might have supported their proliferation and survival across various habitats.

Following closely are **large** and **small [theropods](https://en.wikipedia.org/wiki/Theropoda)**, with 733 and 717 occurrences, respectively. Large theropods, including fearsome predators like Tyrannosaurus and Allosaurus, and small theropods, often agile hunters, underline the ecological roles these carnivores played in controlling other animal populations and possibly influencing the evolutionary pathways of other dinosaurs through predatory pressures.

[Sauropods](https://en.wikipedia.org/wiki/Sauropoda), characterized by their enormous size and long necks, tally 665 occurrences, emphasizing their iconic presence across many Mesozoic landscapes. [Ceratopsians](https://en.wikipedia.org/wiki/Ceratopsia) and [armored dinosaurs](https://en.wikipedia.org/wiki/Thyreophora) with 363 and 307 occurrences, respectively, reflect specialized evolutionary paths with their distinctive horns, frills, and armor, suggesting adaptation to specific environmental and predatory pressures.

### 4. Did dinosaurs get bigger over time?

In [10]:
# Calculate the average age of dinosaurs for better temporal resolution
df['average_age'] = (df['max_ma'] + df['min_ma']) / 2

# Clean the data by dropping rows where length or age data might be missing
df_clean = df.dropna(subset=['length_m', 'average_age'])

# Perform linear regression
slope, intercept, r_value, p_value, std_err = linregress(df_clean['average_age'], df_clean['length_m'])

# Prepare to plot the regression line
df_clean['predicted_length'] = intercept + slope * df_clean['average_age']

# Create the scatter plot and add the regression line
fig = go.Figure()
fig.add_trace(go.Scatter(x=df_clean['average_age'], y=df_clean['length_m'], mode='markers',
                         marker=dict(color='black'), name='Observed Length'))
fig.add_trace(go.Scatter(x=df_clean['average_age'], y=df_clean['predicted_length'], mode='lines',
                         line=dict(color='red'), name='Fit Line'))

# Update layout for clarity and style
fig.update_layout(
    title='Did dinosaurs get bigger over time?',
    xaxis=dict(title='Average Age (million years ago)', autorange='reversed', showgrid=True, gridcolor='lightgrey'),
    yaxis=dict(title='Length (meters)', showgrid=True, gridcolor='lightgrey'),
    plot_bgcolor='white',
    paper_bgcolor='white'
)

fig.show()

# Output the statistical results
print(f"R-squared: {r_value**2:.4f}")
print(f"P-value: {p_value:.4g}")

R-squared: 0.0953
P-value: 1.326e-79


The **R-squared** value indicates the proportion of variance in the dependent variable (dinosaur length) that is predictable from the independent variable (average age). An R-squared of 0.0953 means that about 9.53% of the variability in dinosaur length can be explained by their age. While this value is not particularly high, it suggests that there is a measurable, albeit weak, relationship where dinosaur size tends to vary with age. As an aspiring paleontologist, this could be interpreted to mean that other factors besides age might have a more significant impact on dinosaur size, such as environmental conditions, availability of resources, or evolutionary adaptations specific to different lineages.

The **P-value** is extremely low, far below the common significance level of 0.05, which strongly suggests that the observed relationship between age and length is statistically significant. This means that the probability of obtaining these results by chance, assuming there is no true relationship (null hypothesis), is exceedingly unlikely. This would lead to confidently rejecting the null hypothesis that there is no relationship between the two variables. The relationship, although not explaining a large amount of variance, is statistically reliable.

**From a paleontological perspective**, the findings suggest that while there is a statistically significant correlation between dinosaur age and size, the relationship is relatively weak. The results might hint at complex evolutionary dynamics, with size possibly being influenced by various ecological pressures rather than merely the passage of time.

### 5. An interactive map showing each record.

This interactive Plotly map visualizes the locations of dinosaur fossils. Here's how to use it:

1. **Viewing the Map**: The map is displayed in the center of your notebook. You can see different regions marked with points, each representing a dinosaur fossil.

2. **Interacting with Points**: Hover over a point to see more information about the dinosaur fossil at that location. The information includes the dinosaur's name, type, diet, length, class, family, and region.

3. **Color Coding**: The points are color-coded based on the 'type' of the dinosaur.

4. **Zooming and Panning**: You can zoom in and out of the map using the scroll wheel of your mouse or the zoom buttons on the map. Click and drag the map to pan and explore different regions.

In [19]:
# Replace missing values with 'NaN'
df.fillna('NaN')

# Create an interactive map with Plotly
fig = px.scatter_geo(df, lat='lat', lon='lng', 
                     color='type',
                     title='Dinosaur Fossils Map', 
                     labels={'type': 'Dinosaur type'},
                     color_discrete_sequence=px.colors.qualitative.Dark2,
                     hover_name='name',
                     hover_data={'lat': False, 
                                 'lng': False,  
                                 'type': True, 
                                 'diet': True, 
                                 'length_m': True, 
                                 'class': True, 
                                 'family': True,
                                 'region': True})

# Customize layout
fig.update_geos(projection_type="mollweide", showcoastlines=True, coastlinecolor="white")

fig.update_layout(title_text="Dinosaur Fossils Map", title_x=0.5,
                  width=875, height=500, 
                  template="plotly_dark",
                  geo=dict(showframe=True, showland=True, showcoastlines=True, showcountries=True,
                           countrywidth=1, landcolor='rgb(243, 243, 243)',
                           lakecolor='rgb(255, 255, 255)', 
                           ), 
                  title_font=dict(size=20),
                  showlegend=True)

fig.show()

### 6. Other questions I explored during my analysis

In addition to the main questions, I also explored the following additional questions using the data:

1. Which dinosaurs are the most common and the rarest?
2. Which dinosaur class is the most prevalent?
3. Which dinosaur family is the most prevalent?
4. Is it more common for a larger dinosaur to be an herbivore or a carnivore?
5. What is the average size of dinosaurs within each type category?
6. How many dinosaurs does each geologic period hold?
7. Which region has the highest frequency of dinosaur fossil discoveries?
8. How complete is the dataset in terms of missing values, and what can we do about it?

#### 1. Which dinosaurs are the most common and the rarest?

In [12]:
# Count occurrences of each dinosaur name
name_counts = df['name'].value_counts()

# Get the top 5 most common dinosaurs
top_5_common = name_counts.head(5).items()

# Get the top 5 most rare dinosaurs (those that appear only once are the rarest)
top_5_rare = name_counts[name_counts == 1].head(5).items() if name_counts[name_counts == 1].shape[0] >= 5 else []

def format_results(title, data):
    print("=" * 35)
    print(title)
    print("=" * 35)
    print("Name               | Occurrences")
    print("-" * 35)
    for i, (name, count) in enumerate(data, start=1):
        print(f"{name:<18} | {count}")
    print("=" * 35)

# Output formatted results for common and rare dinosaurs
format_results("Top 5 Most Common Dinosaurs", top_5_common)
if top_5_rare:
    format_results("Top 5 Most Rare Dinosaurs", top_5_rare)
else:
    print("\nNot enough rare dinosaurs to list top 5.")

Top 5 Most Common Dinosaurs
Name               | Occurrences
-----------------------------------
Richardoestesia    | 151
Saurornitholestes  | 136
Triceratops        | 125
Iguanodon          | 111
Troodon            | 111
Top 5 Most Rare Dinosaurs
Name               | Occurrences
-----------------------------------
Qijianglong        | 1
Trinisaura         | 1
Qiyunshanpus       | 1
Ingenia            | 1
Pisanosaurus       | 1


![Iguanodon_galvensis](Iguanodon_galvensis.jpg "Iguanodon galvensis")

_Image of the Iguanodon galvensis from_ [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Iguanodon_galvensis.png)

Starting with the most **common dinosaurs**, we see that Richardoestesia leads the list with 151 occurrences. [Richardoestesia](https://en.wikipedia.org/wiki/Richardoestesia) was a small, carnivorous dinosaur known for its distinctive, blade-like teeth, suggesting a specialized feeding behavior, possibly scavenging, or preying on small vertebrates. [Saurornitholestes](https://en.wikipedia.org/wiki/Saurornitholestes), ranking second with 136 occurrences, was another small theropod, noted for its agility and sharp claws, indicative of a predatory lifestyle. [Triceratops](https://en.wikipedia.org/wiki/Triceratops), found 125 times, stands out as one of the most iconic and recognizable dinosaurs, with its massive skull adorned with three horns and a large frill, traits that likely served defensive functions and intra-species recognition. [Iguanodon](https://en.wikipedia.org/wiki/Iguanodon), with 111 occurrences, was one of the first dinosaurs ever discovered and scientifically described, notable for its distinctive thumb spikes and herbivorous diet. Tied with Iguanodon, [Troodon](https://en.wikipedia.org/wiki/Troodon) features prominently due to its relatively large brain and keen eyesight, characteristics suggesting an elevated level of behavioral sophistication.

![Pisanosaurus](Pisanosaurus.jpg "Pisanosaurus")

_Image of the Pisanosaurus from_ [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Pisanosaurus.jpg)

In stark contrast, the top five most **rare dinosaurs** each have only been found once, highlighting the vast unknowns still prevalent in paleontological research and the potential biases in the fossil record. [Qijianglong](https://en.wikipedia.org/wiki/Qijianglong), a sauropod distinguished by its remarkably long neck, which might have allowed it to forage in large areas or reach high vegetation, is one such rarity. [Trinisaura](https://en.wikipedia.org/wiki/Trinisaura), an Antarctic dinosaur, provides valuable clues about the survival strategies of dinosaurs in polar environments. [Qiyunshanpus](https://www.academia.edu/14230914/Upper_Cretaceous_dinosaur_track_assemblages_and_a_new_theropod_ichnotaxon_from_Anhui_Province_eastern_China), known from only minimal remains, hints at the diversity of small theropods in Asia. Ingenia, also known as [Heyunnia](https://en.wikipedia.org/wiki/Heyuannia), from the late Cretaceous of Mongolia, challenges our understanding with its unique skeletal structure, suggesting rapid evolutionary experiments in small-bodied dinosaurs. Finally, [Pisanosaurus](https://en.wikipedia.org/wiki/Pisanosaurus), an early herbivorous dinosaur from South America, helps shed light on the early evolutionary paths of ornithischian dinosaurs.

#### 2. Which dinosaur class is the most prevalent?

In [13]:
# Count the occurrences of each dinosaur class
class_counts = df['class'].value_counts()

# Get the top 10 most prevalent dinosaur classes
top_classes = class_counts.head()

# Create a bar chart for the top 10 dinosaur classes
fig = px.bar(top_classes, 
             title='Which dinosaur class is the most prevalent?', 
             labels={'index': 'Dinosaur Class', 'value': 'Number of Occurrences'},
             color_discrete_sequence=['black'])

# Update layout for a clean black and white look
fig.update_layout(plot_bgcolor='white', paper_bgcolor='white', showlegend=False)

fig.show()

![Pelvis_Structure-Saurischians_Ornithischians](Pelvis_Structure-Saurischians_Ornithischians.jpg "Pelvis Structure of Saurischians and Ornithischians Dinosaurs")
_Image of Pelvis Structure in Saurischians and Ornithischians Dinosaurs from_ [Britannica](https://cdn.britannica.com/17/93317-050-29267646/Pelvis-structure-saurischians-ornithischians.jpg)

**Saurischia**, the most prevalent class in the data, is distinguished by a "[lizard-hipped](https://www.britannica.com/animal/saurischian)" pelvic structure, with pubis bones pointing downward and forward. This class encompasses two major groups: theropods, which were primarily carnivorous and include famous members like Tyrannosaurus rex and Velociraptor, and sauropodomorphs, which include the largest animals to ever walk the Earth, like Brontosaurus and Diplodocus. The dominance of Saurischians in the fossil record can be attributed to several factors, including their widespread geographic distribution and a wide range of adaptations that allowed them to inhabit diverse ecological niches—from dense forests to arid deserts.

On the other hand, **Ornithischia**, or "[bird-hipped](https://www.britannica.com/animal/ornithischian)" dinosaurs, despite their name, are not the direct ancestors of birds. Their pelvic structure resembles that of birds, with both pubis and ischium pointing backward, parallel to each other. This group includes a variety of herbivorous dinosaurs such as Triceratops, Stegosaurus, and the duck-billed Hadrosaurs. Ornithischians were known for their varied defensive adaptations, including armor plating, horns, and dental specializations for processing plant material, which indicates a highly competitive and diverse herbivorous niche.

#### 3. Which dinosaur family is the most prevalent?

In [14]:
# Count the occurrences of each dinosaur class
class_counts = df['family'].value_counts()

# Get the top 10 most prevalent dinosaur classes
top_family = class_counts.head(10)

# Create a bar chart for the top 10 dinosaur classes
fig = px.bar(top_family,
             orientation='h',
             title='Which dinosaur family is the most prevalent?', 
             labels={'index': 'Dinosaur Family', 'value': 'Number of Occurrences'},
             color_discrete_sequence=['black'])

# Update layout for a clean black and white look
fig.update_layout(plot_bgcolor='white', paper_bgcolor='white', yaxis={'categoryorder': 'total ascending'}, showlegend=False)

fig.show()

Leading the list is [Dromaeosauridae](https://en.wikipedia.org/wiki/Dromaeosauridae), with 450 occurrences, famously known for members like Velociraptor. Dromaeosaurids are often characterized by their sickle-shaped claws and keen predatory instincts, indicative of their role as agile hunters. The [Hadrosauridae](https://en.wikipedia.org/wiki/Hadrosauridae), or duck-billed dinosaurs, follow with 319 occurrences. Hadrosaurs were extraordinarily successful herbivores, known for their sophisticated chewing mechanisms and often elaborate cranial ornamentations, which likely played roles in social behavior and species recognition.

![Triceratops](Triceratops.jpg "Triceratops")

_Image of Triceratops by Leonello Calvett from_ [Alamy](https://www.alamy.com/stock-photo-triceratops-dinosaur-very-well-detailed-and-scientifically-correct-47577828.html)

The [Ceratopsidae](https://en.wikipedia.org/wiki/Ceratopsidae), with 287 findings, were another group of herbivorous dinosaurs noted for their impressive head frills and arrays of facial horns, which served as defensive tools and as signals within their species. [Troodontidae](https://en.wikipedia.org/wiki/Troodontidae), with 245 occurrences, were small to medium-sized, bird-like dinosaurs known for their large brains and keen senses, suggesting high intelligence and complex behaviors. [Tyrannosauridae](https://en.wikipedia.org/wiki/Tyrannosauridae), totaling 183 occurrences, includes the iconic Tyrannosaurus rex and represents some of the largest known predatory dinosaurs, characterized by massive skulls and powerful jaws capable of delivering fatal bites to their prey.

![Tyrannosaurus_rex](Tyrannosaurus_rex.jpg "Tyrannosaurus rex")

_Image of Tyrannosaurus rex by Leonello Calvetti from_ [Alamy](https://www.alamy.com/tyrannosaurus-rex-dinosaur-full-body-photo-realistic-representation-image66152351.html)

[Diplodocidae](https://en.wikipedia.org/wiki/Diplodocidae), found 162 times, were massive sauropods like Diplodocus and Apatosaurus, renowned for their long necks and tails, which they used to reach high vegetation and possibly as defensive whips. [Stegosauridae](https://en.wikipedia.org/wiki/Stegosauridae), with 120 occurrences, included dinosaurs such as Stegosaurus, with their distinctive rows of back plates and tail spikes, adaptations that likely served defensive and thermoregulatory purposes. 

The [Nodosauridae](https://en.wikipedia.org/wiki/Nodosauridae), recording 118 findings, were similar to ankylosaurs but lacked tail clubs; instead, they relied on heavy, bony armor embedded in their skin for protection. [Ankylosauridae](https://en.wikipedia.org/wiki/Ankylosauridae), with 108 occurrences, were built like tanks, featuring extensive armor and a characteristic club at the end of their tails, making them well-equipped to defend against predators. Lastly, the [Allosauridae](https://en.wikipedia.org/wiki/Allosauridae), with 104 occurrences, were formidable predators, slightly smaller than tyrannosaurs but just as fierce, with adaptations for hunting large prey, including sharp teeth and powerful forelimbs.

#### 4. Is it more common for a larger dinosaur to be an herbivore or a carnivore?

In [15]:
# Drop rows where either diet or length_m is missing
df_clean = df.dropna(subset=['diet', 'length_m'])

# Group by 'diet' and calculate the average 'length_m'
average_size_by_diet = df_clean.groupby('diet')['length_m'].mean().reset_index()

# Sort results for better visualization
average_size_by_diet = average_size_by_diet.sort_values(by='length_m', ascending=False)

# Create a bar chart to visualize the average sizes
fig = px.bar(average_size_by_diet,
             x='diet',
             y='length_m',
             title='Average Size of Dinosaurs by Diet',
             labels={'length_m': 'Average Length (meters)', 'diet': 'Diet'},
             color_discrete_sequence=['black'])

# Update layout for a clean black and white look
fig.update_layout(plot_bgcolor='white', paper_bgcolor='white', showlegend=False)

fig.show()

On average, **herbivorous** dinosaurs appear to be the largest, with an average length of **10.29 meters**. This is likely due to the nature of their diet. Herbivorous dinosaurs, such as the well-known [Apatosaurus](https://en.wikipedia.org/wiki/Apatosaurus) and [Brachiosaurus](https://en.wikipedia.org/wiki/Brachiosaurus), consumed copious quantities of plant matter. The digestion of plant material requires a long digestive tract and a large body size to accommodate this. Additionally, a larger size could have allowed these dinosaurs to reach vegetation that smaller herbivores could not, giving them a competitive advantage.

**Carnivorous** dinosaurs, on the other hand, had an average length of **5.60 meters**. Carnivores, like the [Tyrannosaurus rex](https://en.wikipedia.org/wiki/Tyrannosaurus) or [Velociraptor](https://en.wikipedia.org/wiki/Velociraptor), typically have a more streamlined body designed for active hunting. Their size would have been a balance between being large enough to overpower their prey, but small enough to maintain the speed and agility necessary for hunting.

![Velociraptor_restraining_an_oviraptor](Velociraptor_restraining_an_oviraptor.jpg)

_The dromaeosaurid Velociraptor mongoliensis is depicted restraining a juvenile oviraptorosaur using the mantling technique, which has been proposed for dromaeosaurids and is done by modern birds of prey. Image from_ [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Velociraptor_restraining_an_oviraptorosaur_by_durbed.jpg)

The smallest of the three groups are the **omnivorous** dinosaurs, with an average length of **4.21 meters**. Omnivores, such as the [Oviraptor](https://en.wikipedia.org/wiki/Oviraptor), had a varied diet that could include small animals, eggs, and plant matter. Their smaller size might reflect the broad range of food sources they exploited, including smaller prey that larger carnivores might ignore, and a lifestyle that required a high degree of versatility rather than specialization.

#### 5. What is the average size of dinosaurs within each type category?

In [16]:
# Drop rows where the type or length_m is missing
df_clean = df.dropna(subset=['type', 'length_m'])

# Group by 'type' and calculate the average 'length_m'
average_size_by_type = df_clean.groupby('type')['length_m'].mean().reset_index()

# Sort results for better visualization
average_size_by_type = average_size_by_type.sort_values(by='length_m', ascending=False)

# Create a bar chart to visualize the average sizes
fig = px.bar(average_size_by_type,
             x='type',
             y='length_m',
             title='Average Size of Dinosaurs by Type',
             labels={'length_m': 'Average Length (meters)', 'type': 'Dinosaur Type'},
             color_discrete_sequence=['black'])

# Update layout for a clean look
fig.update_layout(plot_bgcolor='white', paper_bgcolor='white', showlegend=False)

fig.show()

![Brachiosaurus](Brachiosaurus.jpg "Brachiosaurus")

_Image of the Brachiosaurus from_ [Wikimedia Commons](https://en.wikipedia.org/wiki/File:Brachiosaurus_altithorax_side_profile.png)

At the top of the size spectrum are the **sauropods**, with an impressive average length of 17.84 meters. Sauropods, such as [Apatosaurus](https://en.wikipedia.org/wiki/Apatosaurus) and [Brachiosaurus](https://en.wikipedia.org/wiki/Brachiosaurus), are renowned for their enormous bodies, long necks, and tails. These massive herbivores likely used their size as a primary adaptation to deter predators and to reach high vegetation that other herbivores could not, thereby accessing a feeding niche unavailable to others. Their enormous size also supported a voluminous digestive system needed to process substantial amounts of plant material; a diet that provided the energy necessary to sustain their monumental bodies.

Following sauropods are the **large theropods**, averaging 8.86 meters in length. This group includes formidable predators like [Tyrannosaurus rex](https://en.wikipedia.org/wiki/Tyrannosaurus) and [Allosaurus](https://en.wikipedia.org/wiki/Allosaurus). The size of these theropods played a crucial role in their hunting strategies, allowing them to overpower large prey and dominate their ecosystems. Despite their size, these theropods were likely quite agile, capable of quick bursts of speed essential for catching prey.

**Ornithopods**, with an average length of 7.14 meters, represent a diverse group of mostly herbivorous dinosaurs, including duck-billed [Hadrosaurs](https://en.wikipedia.org/wiki/Hadrosauridae). Their size varied widely, but generally, ornithopods developed sophisticated chewing mechanisms and often complex social structures, which might have been facilitated by their larger size in terms of defense and foraging efficiency.

![Ankylosaurus](Ankylosaurus.jpg "Ankylosaurus")

_Image of Ankylosaurus by Catmando from_ [Shuttersctock](https://www.shutterstock.com/image-illustration/ankylosaurus-01-armored-dinosaur-creataceous-period-85156309)

The **armored dinosaurs**, averaging 6.28 meters, include the [Ankylosaurus](https://en.wikipedia.org/wiki/Ankylosaurus) and [Stegosaurus](https://en.wikipedia.org/wiki/Stegosaurus). Their robust bodies were adorned with defensive features such as bony plates, spikes, and in some cases, clubbed tails, which likely compensated for their lack of speed with increased defensive capabilities against predators. [Ceratopsians](https://en.wikipedia.org/wiki/Ceratopsia), at an average length of 5.69 meters, were another group of herbivores notable for their distinctive skull frills and facial horns. These features not only served protective functions but also likely played roles in social signaling and species recognition within their environments.

Finally, the **small theropods**, averaging 1.89 meters, such as [Velociraptor](https://en.wikipedia.org/wiki/Velociraptor), demonstrate that not all theropods were giant predators. These smaller dinosaurs were likely highly agile and used their size to navigate dense vegetation and exploit smaller prey or scavenging opportunities, reflecting a different ecological niche from their larger counterparts.

#### 6. How many dinosaurs does each geologic period hold?

In [17]:
# Define geologic periods
periods = {
    'Triassic': (251.902, 201.4),
    'Jurassic': (201.4, 145),
    'Cretaceous': (145, 66)
}

# Function to determine the period of a dinosaur based on max_ma and min_ma
def assign_period(row):
    for period, (start, end) in periods.items():
        if row['max_ma'] <= start and row['min_ma'] >= end:
            return period
    return 'Unknown'

# Apply the function to each row in the DataFrame
df['Period'] = df.apply(assign_period, axis=1)

# Drop rows where the period is 'Unknown'
df = df[df['Period'] != 'Unknown']

# Count occurrences in each period
period_counts = df['Period'].value_counts()

# Create a bar chart to visualize the number of dinosaurs in each period
fig = px.bar(period_counts, 
             title='How many dinosaurs does each geologic period hold?',
             labels={'index': 'Geologic Period', 'value': 'Count of Dinosaurs'},
             color_discrete_sequence=['black'])

# Update layout for clarity
fig.update_layout(plot_bgcolor='white', paper_bgcolor='white', showlegend=False)

fig.show()

![The_Triassic_Period](The_Triassic_Period.jpg "An artist's depiction captures the burst of new life that occurred in the early- to mid-Triassic period. Massive extinctions at the end of the preceding Permian period allowed the plants and animals that survived to grow and diversify relatively free of competition and predators. These conditions gave rise to the dinosaurs, pterosaurs, and early crocodilians.")

_An artist's depiction of the Triassic Period. Image from_ [National Geographic](https://i.natgeofe.com/n/4ada967a-e4eb-4b53-b9f6-77958d7a6e2d/1256_4x3.jpg)

The [Triassic period](https://australian.museum/learn/australia-over-time/evolving-landscape/the-triassic-period-252-201-million-years-ago/) (252 to 201 million years ago), with 139 recorded dinosaur species, marks the earliest phase of dinosaur evolution. During this time, dinosaurs were just beginning to emerge and differentiate from other archosaurian reptiles. The relatively low diversity of dinosaurs in the Triassic suggests that they were initially few in number and variety, likely occupying limited ecological niches as they competed with other dominant reptilian groups like the Pseudosuchians. However, the end of the Triassic witnessed a mass extinction event, which, while devastating many other groups, provided an opportunity for dinosaurs to expand and fill vacated ecological niches.


![The_Jurassic_Period](The_Jurassic_Period.jpg "The Jurassic period (199.6 million to 145.5 million years ago) was characterized by a warm, wet climate that gave rise to lush vegetation and abundant life. Many new dinosaurs emerged—in great numbers. Among them were stegosaurs, brachiosaurs, allosaurs, and many others.")

_An artist's depiction of the Jurassic Period. Image from_ [National Geographic](https://i.natgeofe.com/n/b84387cb-28f5-4611-a221-65ae275ae2e3/1028_4x3.jpg)

Transitioning into the [Jurassic period](https://australian.museum/learn/australia-over-time/evolving-landscape/the-jurassic-period-201-145-million-years-ago/) (201 to 145 million years ago), we observe a significant increase in dinosaur diversity, with 1251 species documented. This period is often considered the "Age of the Dinosaurs" where they became the dominant terrestrial vertebrates. The Jurassic marks a period of robust evolutionary experimentation and expansion, with the appearance of iconic groups such as the large sauropods and the first birds. The diversification of plant life during the Jurassic, including the spread of gymnosperms and the emergence of the first angiosperms, likely supported the evolution of a variety of herbivorous dinosaurs, which in turn supported a greater diversity of carnivorous species.

![The_Cretaceous_Period](The_Cretaceous_Period.jpg "This rendering of Cretaceous life shows the diverse range of dinosaurs that lived between 145 and 65 million years ago, including maiasaurs (front left); tarbosaurs (top right), and pterosaurs (top center). In the foreground are depicted the first flowering plants and one of the earliest mammal relatives, both of which developed during this period and went on to survive the dinosaur extinction at the end of the Cretaceous.")

_An artist's depiction of the Cretaceous Period. Image from_ [National Geographic](https://i.natgeofe.com/n/9c56eafe-8285-4eef-b0d4-e12473098742/907_4x3.jpg)

The [Cretaceous period](https://australian.museum/learn/australia-over-time/evolving-landscape/the-cretaceous-period/) (145 to 66 million years ago) highlights the peak of dinosaur diversity with 3459 species, underscoring a complex, well-established dinosaurian world. This period was characterized by significant geological and climatic changes, including increased tectonic activity and a warmer climate, which led to high sea levels and the formation of new ecological niches. The Cretaceous is notable for the flowering plants' proliferation, which revolutionized food sources and habitats, further influencing dinosaur evolution. Additionally, this period featured the further evolution and diversification of both herbivorous and carnivorous dinosaurs, culminating in some of the most sophisticated and specialized forms, such as the armored ankylosaurs and the horned ceratopsians.

#### 7. Which region has the highest frequency of dinosaur fossil discoveries?

In [18]:
# Count the number of dinosaur occurrences by region and get the top 10
top_regions = df['region'].value_counts().head(10)

# Create a bar chart for the top 10 regions using a black and white color scheme
fig = px.bar(top_regions,
             orientation='h',
             title='Which region has the highest frequency of dinosaur fossil discoveries?', 
             labels={'index': 'Region', 'value': 'Number of Occurrences'}
            )

# Update layout for a clean black and white look
fig.update_layout(plot_bgcolor='white', paper_bgcolor='white', yaxis={'categoryorder': 'total ascending'}, showlegend=False)
fig.update_traces(marker_color='black', marker_line_color='black', marker_line_width=1.5) 

fig.show()

![Dinosaur_Provincial_Park](Dinosaur_Provincial_Park.jpg "The Dinosaur Provincial Park in Alberta, Canada")

_The Dinosaur Provincial Park in Alberta, Canada. Image from_ [Alberta Parks](https://www.albertaparks.ca/media/6495772/dino-ambassador_kelsy-nielson7.jpg)

**Alberta**, Canada leads the list with 650 occurrences, highlighting the region's rich paleontological sites, particularly the famous [Dinosaur Provincial Park](https://en.wikipedia.org/wiki/Dinosaur_Provincial_Park), part of the larger Western Canadian Sedimentary Basin. This area is renowned for its exceptional preservation conditions and the abundance of Late Cretaceous deposits, which have provided a wealth of dinosaur fossils.

**Wyoming** and **Montana**, with 514 and 445 occurrences respectively, are part of the [Morrison Formation](https://en.wikipedia.org/wiki/Morrison_Formation), a sequence of Upper Jurassic sedimentary rock known for being one of the most fruitful sources of dinosaur fossils in North America. The variety and quantity of Jurassic-age fossils found here have been crucial in understanding the diversity and ecology of late Jurassic ecosystems.

**Omnogovi** in Mongolia, with 206 occurrences, is notable for its contributions to our understanding of dinosaurs from the Cretaceous period. The region's [Gobi Desert](https://en.wikipedia.org/wiki/Gobi_Desert) has yielded numerous well-preserved fossils, providing insights into dinosaur species that inhabited Asia, including many unique forms not found elsewhere.

---
![Jurassic_Coast](Jurassic_Coast.jpg "The Jurassic Coast in Dorset, looking west from St. Aldhelm's Head. Visible coastal features from right to left are: Chapman's Pool (the nearest bay in the foreground), Egmont Point, Egmont Bight, Swyre Head, the Kimmeridge Ledges, Gad Cliff, Worbarrow Bay (below the white cliffs), and Bindon Hill.")

_The Jurassic Coast west of St. Aldhelm's Head. Image from_ [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Jurassic_Coast_west_of_St_Aldhelm%27s_Head.JPG)

**England**, with 200 occurrences, has a long history of dinosaur paleontology, being one of the first places where dinosaur fossils were systematically studied. The [Jurassic Coast](https://en.wikipedia.org/wiki/Jurassic_Coast) and the [Wealden Clay Formations](https://en.wikipedia.org/wiki/Weald_Clay) have been particularly prolific, offering fossils that have shed light on the early diversity of dinosaurs during the Jurassic and Cretaceous periods.

---
![Morrison_Formation](Morrison_Formation.jpg "The distinctive banding of the Morrison Formation, a group of rock layers that occur throughout Dinosaur National Monument. The formation originated as muds and sands laid down by ancient rivers, and some of its outcrops have been found to contain 150-million-year-old dinosaur fossils like those found at the monument's Dinosaur Quarry.")

_The Morrison Formation. Image from_ [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Green_River_UT_2005-10-14_2104.jpg)

**Colorado**, **New Mexico**, and **Utah**, with 187, 164, and 138 occurrences respectively, also benefit from being part of the Morrison Formation and similar Mesozoic sedimentary deposits. These regions have contributed significantly to the North American dinosaur fossil record, providing a plethora of specimens that have helped paleontologists piece together the continent's ancient faunal history.

**Lindi** in Tanzania, with 111 occurrences, and **Liaoning** in China, with 80 occurrences, represent important paleontological sites outside of North America. Lindi is part of the rich [East African Rift system](https://phys.org/news/2018-10-paleontologists-species-east-african-rift.html), which has provided fossils that offer a rare glimpse into the dinosaur faunas of ancient Gondwana during the Jurassic and Cretaceous periods. [Liaoning](https://www.amnh.org/exhibitions/dinosaurs-ancient-fossils/liaoning-diorama/fossils-from-liaoning) is renowned for its exceptionally well-preserved feathered dinosaur fossils, which have revolutionized our understanding of dinosaur plumage and bird evolution.

#### 8. How complete is the dataset in terms of missing values, and what can we do about it?

In [19]:
# Calculate the number of missing values for each column
missing_values = df.isnull().sum()
print("Missing values in each column:")
print(missing_values)

Missing values in each column:
occurrence_no       0
name                0
diet             1311
type             1311
length_m         1339
max_ma              0
min_ma              0
region             40
lng                 0
lat                 0
class               0
family           1398
average_age         0
Period              0
dtype: int64


Columns such as `occurrence_no`, `name`, `max_ma`, `min_ma`, `lng`, `lat`, and `class` are complete and reliable for analysis, indicating robust recording of key identifiers, geographic coordinates, and temporal data.

However, significant gaps exist in other critical areas:
1. `diet` and `type`: Both columns have 1311 missing values, suggesting a dependency in data collection. This lack impedes detailed ecological and evolutionary studies.
2. `length`: With 1339 missing entries, this column's data gaps limit morphological and ecological analyses.
3. `family`: The most significant gaps, with 1398 missing values, obstruct finer taxonomic classifications and phylogenetic studies.
4. `region`: Missing in 40 cases, this affects biogeographic studies and interpretations of distribution patterns and migration histories.

Addressing these missing values is crucial for comprehensive analysis. Imputation techniques or excluding certain analyses where missing data significantly impact results may be necessary. Understanding the reasons behind these gaps—whether due to data rarity, measurement difficulties, or preservation limitations—will also guide future data collection efforts and improve findings interpretation.

## 1.5. Conclusion

The analysis of the dinosaur dataset has provided a rich tableau of insights into the Mesozoic era, revealing not only the vast diversity and distribution of dinosaurs but also their evolutionary trends and ecological adaptations. With 1,042 unique dinosaur names recorded, the dataset underscores the immense biodiversity that once thrived, reflecting a wide array of ecological niches across various continents. The presence of giants such as Supersaurus and Argentinosaurus, which span up to 35 meters, exemplifies the peak physical adaptations achieved by sauropods, highlighting their dominance in size among their contemporaries. Furthermore, the predominance of ornithopods in the dataset, with 811 occurrences, illustrates their successful adaptation strategies, possibly driven by their dietary flexibility and social behaviors, which might have contributed significantly to their proliferation and survival.

Delving deeper, the analysis has also shed light on the evolutionary dimensions of these ancient giants, with findings suggesting a weak but statistically significant relationship between dinosaur sizes and their ages, hinting that factors other than mere chronological progression—such as environmental changes and resource availability—played crucial roles in their physical development. The regional analysis revealing Alberta and Wyoming as hotspots for dinosaur fossil discoveries emphasizes the geological and environmental idiosyncrasies that have preserved these windows into the past. However, the substantial number of missing data, especially in critical fields like diet, type, and length, poses significant challenges, highlighting the inherent gaps in our paleontological records. These gaps underscore the necessity for continued and focused excavation efforts, improved data recording practices, and advanced analytical techniques to fill in the missing pieces of the prehistoric puzzle, thereby refining our understanding of dinosaur evolution and their ecological legacies.

## 1.6. Recommendations

Based on the insights garnered from the analysis of the dinosaur dataset, the following recommendations are proposed to enhance future research and data management:

1. **Enhanced Data Collection Methods**: Implement standardized protocols for data entry to minimize inconsistencies and missing data, especially in key fields like diet, type, and length. This could involve training sessions for field researchers and data entry personnel on the importance of comprehensive data collection.

2. **Advanced Analytical Techniques**: Employ more sophisticated statistical methods and machine learning algorithms to analyze the existing data. This could help uncover patterns and trends that are not immediately apparent, providing deeper insights into the evolutionary history and ecological dynamics of dinosaurs.

3. **Interdisciplinary Collaboration**: Encourage collaboration between paleontologists, climatologists, and geologists to better understand the environmental factors influencing dinosaur evolution and distribution. This could include joint research projects and shared data repositories.

4. **Public Engagement and Education**: Develop interactive displays and digital resources based on the dataset to engage the public and educational institutions. This not only raises awareness about dinosaur science but also encourages public support for paleontological research.

5. **Continued Funding and Support**: Advocate for increased funding for paleontological research from government bodies, educational institutions, and private donors. Sustained financial support is crucial for long-term research projects, particularly those involving extensive fieldwork and advanced technological applications.

6. **Data Sharing Initiatives**: Promote the creation of an open-access database where researchers worldwide can share their findings and data on dinosaur fossils. This would enhance the global understanding of dinosaur biodiversity and foster a collaborative research environment.

## 1.7. Links/References

1. Allosauridae. (2024, April 27). Wikipedia. https://en.wikipedia.org/wiki/Allosauridae
2. Allosaurus. (2024, April 27). Wikipedia. https://en.wikipedia.org/wiki/Allosaurus
3. Ankylosauridae. (2024, April 23). Wikipedia. https://en.wikipedia.org/wiki/Ankylosauridae
4. Ankylosaurus. (2024, March 15). Wikipedia. https://en.wikipedia.org/wiki/Ankylosaurus
5. Apatosaurus. (2024, April 10). Wikipedia. https://en.wikipedia.org/wiki/Apatosaurus
6. Argentinosaurus. (2024, April 25). Wikipedia. https://en.wikipedia.org/wiki/Argentinosaurus
7. Argentinosaurus. (2024, April 25). Wikipedia. https://en.wikipedia.org/wiki/Argentinosaurus#/media/File:Argentinosaurus_BW.jpg
8. Brachiosaurus. (2024, April 1). Wikipedia. https://en.wikipedia.org/wiki/Brachiosaurus
9. Ceratopsia. (2024, April 23). Wikipedia. https://en.wikipedia.org/wiki/Ceratopsia
10. Ceratopsidae. (2024, March 31). Wikipedia. https://en.wikipedia.org/wiki/Ceratopsidae
11. Dinosaur Provincial Park. (2024, February 26). Wikipedia. https://en.wikipedia.org/wiki/Dinosaur_Provincial_Park
12. Diplodocidae. (2024, February 11). Wikipedia. https://en.wikipedia.org/wiki/Diplodocidae
13. Dromaeosauridae. (2004, May 5). Wikipedia. https://en.wikipedia.org/wiki/Dromaeosauridae
14. Fossils from Liaoning | AMNH. (n.d.). American Museum of Natural History. https://www.amnh.org/exhibitions/dinosaurs-ancient-fossils/liaoning-diorama/fossils-from-liaoning
15. GEOL 104 Ornithopoda: Beaks, bills & crests. (n.d.). https://www.geol.umd.edu/~tholtz/G104/lectures/104ornithop.html
16. Gobi Desert. (2024, March 28). Wikipedia. https://en.wikipedia.org/wiki/Gobi_Desert
17. Hadrosauridae. (2024, April 24). Wikipedia. https://en.wikipedia.org/wiki/Hadrosauridae
18. Heyuannia. (2024, January 27). Wikipedia. https://en.wikipedia.org/wiki/Heyuannia
19. Iguanodon. (2024, April 23). Wikipedia. https://en.wikipedia.org/wiki/Iguanodon
20. Jurassic Coast. (2024, February 7). Wikipedia. https://en.wikipedia.org/wiki/Jurassic_Coast
21. Klein, H. (2015, July 20). Upper Cretaceous dinosaur track assemblages and a new theropod ichnotaxon from Anhui Province, eastern China. www.academia.edu. https://www.academia.edu/14230914/Upper_Cretaceous_dinosaur_track_assemblages_and_a_new_theropod_ichnotaxon_from_Anhui_Province_eastern_China
22. Morrison Formation. (2024, March 24). Wikipedia. https://en.wikipedia.org/wiki/Morrison_Formation
23. Nodosauridae. (2023, November 30). Wikipedia. https://en.wikipedia.org/wiki/Nodosauridae
24. Ornithischian | Triassic, Jurassic, Cretaceous. (1998, July 20). Encyclopedia Britannica. https://www.britannica.com/animal/ornithischian
25. Oviraptor. (2024, April 15). Wikipedia. https://en.wikipedia.org/wiki/Oviraptor
26. Pisanosaurus. (2024, April 12). Wikipedia. https://en.wikipedia.org/wiki/Pisanosaurus
27. Qijianglong. (2024, March 16). Wikipedia. https://en.wikipedia.org/wiki/Qijianglong
28. Richardoestesia. (2024, April 21). Wikipedia. https://en.wikipedia.org/wiki/Richardoestesia
29. Saurischian | Triassic, Jurassic & Cretaceous Species. (1998, July 20). Encyclopedia Britannica. https://www.britannica.com/animal/saurischian
30. Sauropoda. (2024, April 3). Wikipedia. https://en.wikipedia.org/wiki/Sauropoda
31. Saurornitholestes. (2024, April 8). Wikipedia. https://en.wikipedia.org/wiki/Saurornitholestes
32. Smithsonian Institution. (n.d.). Image Cover. Retrieved April 28, 2024, from https://www.si.edu/sites/default/files/newsdesk/press_releases/nhb2017-00021.jpg
33. Stegosauridae. (2024, April 5). Wikipedia. https://en.wikipedia.org/wiki/Stegosauridae
34. Stegosaurus. (2024, April 3). Wikipedia. https://en.wikipedia.org/wiki/Stegosaurus
35. Supersaurus. (2024, March 30). Wikipedia. https://en.wikipedia.org/wiki/Supersaurus
36. Supersaurus. (2024, March 30). Wikipedia. https://en.wikipedia.org/wiki/Supersaurus#/media/File:Supersaurus_dinosaur.png
37. Tang, C. M. (2024, March 8). Mesozoic Era | Geology, Climate & Lifeforms. Encyclopedia Britannica. https://www.britannica.com/science/Mesozoic-Era
38. The Cretaceous Period (146-65 million years ago). (n.d.). The Australian Museum. https://australian.museum/learn/australia-over-time/evolving-landscape/the-cretaceous-period/
39. The Jurassic Period (201 - 145 million years ago). (n.d.). The Australian Museum. https://australian.museum/learn/australia-over-time/evolving-landscape/the-jurassic-period-201-145-million-years-ago/
40. The Paleobiology Database. (n.d.). https://paleobiodb.org/#/
41. The Triassic Period (252 - 201 million years ago). (n.d.). The Australian Museum. https://australian.museum/learn/australia-over-time/evolving-landscape/the-triassic-period-252-201-million-years-ago/
42. Theropoda. (2003, August 20). Wikipedia. https://en.wikipedia.org/wiki/Theropoda
43. Thyreophora. (2024, April 15). Wikipedia. https://en.wikipedia.org/wiki/Thyreophora
44. Triceratops. (2024, April 21). Wikipedia. https://en.wikipedia.org/wiki/Triceratops
45. Troodon. (2024, April 13). Wikipedia. https://en.wikipedia.org/wiki/Troodon
46. Tyrannosauridae. (2024, March 16). Wikipedia. https://en.wikipedia.org/wiki/Tyrannosauridae
47. Tyrannosaurus. (2024, April 24). Wikipedia. https://en.wikipedia.org/wiki/Tyrannosaurus
48. Velociraptor. (2024, April 21). Wikipedia. https://en.wikipedia.org/wiki/Velociraptor
49. Weald Clay. (2024, March 23). Wikipedia. https://en.wikipedia.org/wiki/Weald_Clay
50. X, S. (2018, October 30). Paleontologists discovered six new species in the East African Rift. https://phys.org/news/2018-10-paleontologists-species-east-african-rift.html

# Part 2: Recapturing Movies 🎥 

![Cover_Image_2](Cover_Image_2.jpg "The Iconic Jaws Movie Poster")

_The Iconic "Jaws" Movie Poster. Cover Image from_ [All The Right Movies](https://alltherightmovies.com/feature/jaws-the-creation-of-hollywoods-most-iconic-movie-poster/)

## Key Findings

- The database contains a comprehensive catalog of 4,968 films, offering a robust base for identifying historical trends and patterns in film production and success.

- The U.S. dominates production with 3,750 films, followed by the U.K. (443), France (153), Canada (123), and Germany (97), showcasing these nations' influence on global cinema.

- Budgets have evolved dramatically, from a modest \$385,907 in 1916 to an average of \$47.65 million in the 2000s, aligning with technological progress and larger-scale production.

- Gross revenue per film surged from \$1.46 million in the 1920s to an average of \$58.73 million in the 2010s, shaped by blockbuster franchises and global distribution.

- Out of 3,892 films with complete financial data, 2,051 (52.7%) were profitable, demonstrating a consistent opportunity for financial success amid inherent industry risks.

- The highest earners include Star Wars: Episode IV - A New Hope (1977), E.T. the Extra-Terrestrial (1982), and The Lion King (1994), primarily from the U.S., emphasizing the influence of Hollywood.

- The Texas Chain Saw Massacre (1974) led the ROI chart with 36,800%, followed by American Graffiti (1973) and Rocky (1976), exemplifying the potential profitability of high-impact, low-budget storytelling.

## 2.1. Overview

I have recently been hired by a large movie studio to perform data analysis aimed at guiding the development of new films that encapsulate the charm and appeal of classic cinema. Our objective is to delve into historical data to identify the elements that contributed to the success of films released before 1997, as exemplified by iconic hits such as "Titanic." By understanding these patterns, we aim to generate ideas that could lead to future blockbuster successes.

**The objectives of this analysis includes:**

- Determine the total number of movies available.  
- Evaluate the completeness of key financial metrics (gross and budget) and recommend strategies for handling any missing data.  
- Identify the diversity of movie certifications present in the dataset, which could indicate the target audience demographic. 
- Ascertain the top five countries in terms of movie production, providing insight into major film-producing regions before 1997.  
- Compare the average duration of movies based on their language, focusing on English and French films to see if there’s a significant difference in their typical lengths.
- Conduct an in-depth analysis of the dataset to identify and present any additional important findings found during the analysis.

## 2.2. Data Description

The data was collected from [IMDb](https://www.imdb.com/). The following table gives some descriptions of our column names or key variables from the `cinema.films` table::

| Column name | Description |
|---|---|
| id | Unique movie identifier. |
| title | The title of the movie. |
| release_year | The year the movie was released to the public. |
| country | The country in which the movie was released. |
| duration | The runtime of the movie, in minutes. |
| language | The original language the movie was produced in. |
| certification | The [rating](https://simple.wikipedia.org/wiki/Motion_Picture_Association_film_rating_system) the movie was given based on their suitability for audiences. |
| gross | The revenue the movie generated at the box office, in USD. |
| budget | The available budget the production had for producing the movie, in USD. |

## 2.3. Initial Data Exploration

In this second part of our notebook, we focus on analyzing our movie dataset using SQL. We begin by taking a peek at our table to understand its structure and contents, check for any potential issues, such as missing values, and derive basic summary statistics that will inform our subsequent analysis.

In [20]:
-- Retrieve all columns and rows from the films table
SELECT * FROM cinema.films;

id  ...     budget
0        1  ...   385907.0
1        2  ...   100000.0
2        3  ...   245000.0
3        4  ...  6000000.0
4        5  ...        NaN
...    ...  ...        ...
4963  4964  ...        NaN
4964  4965  ...        NaN
4965  4966  ...        NaN
4966  4967  ...        NaN
4967  4968  ...        NaN

[4968 rows x 9 columns]

In [21]:
-- Retrieve the data types of the columns in the films table
SELECT 
    pg_typeof(id) AS id_type,
    pg_typeof(title) AS title_type,
    pg_typeof(release_year) AS release_year_type,
    pg_typeof(country) AS country_type,
    pg_typeof(duration) AS duration_type,
    pg_typeof(language) AS language_type,
    pg_typeof(certification) AS certification_type,
    pg_typeof(gross) AS gross_type,
    pg_typeof(budget) AS budget_type
FROM cinema.films
LIMIT 1;

id_type         title_type  ... gross_type budget_type
0  integer  character varying  ...     bigint      bigint

[1 rows x 9 columns]

In [22]:
-- Check for null values in each column of the films table
SELECT
	COUNT(*) AS total_records,
	SUM(CASE WHEN id IS NULL THEN 1 ELSE 0 END) AS null_id,
    SUM(CASE WHEN title IS NULL THEN 1 ELSE 0 END) AS null_title,
    SUM(CASE WHEN release_year IS NULL THEN 1 ELSE 0 END) AS null_release_year,
    SUM(CASE WHEN country IS NULL THEN 1 ELSE 0 END) AS null_country,
    SUM(CASE WHEN duration IS NULL THEN 1 ELSE 0 END) AS null_duration,
    SUM(CASE WHEN language IS NULL THEN 1 ELSE 0 END) AS null_language,
    SUM(CASE WHEN certification IS NULL THEN 1 ELSE 0 END) AS null_certification,
    SUM(CASE WHEN gross IS NULL THEN 1 ELSE 0 END) AS null_gross,
    SUM(CASE WHEN budget IS NULL THEN 1 ELSE 0 END) AS null_budget
FROM cinema.films;

total_records  null_id  ...  null_gross  null_budget
0           4968        0  ...         810          430

[1 rows x 10 columns]

In [23]:
-- Retrieve summary statistics for duration, gross, and budget, along with counts of distinct countries and languages in the films table
SELECT
    AVG(duration) AS average_duration,
    MIN(duration) AS minimum_duration,
    MAX(duration) AS maximum_duration,
    AVG(gross) AS average_gross,
    MIN(gross) AS minimum_gross,
    MAX(gross) AS maximum_gross,
    AVG(budget) AS average_budget,
    MIN(budget) AS minimum_budget,
    MAX(budget) AS maximum_budget,
    COUNT(DISTINCT country) AS distinct_countries,
    COUNT(DISTINCT language) AS distinct_languages
FROM cinema.films;

average_duration  minimum_duration  ...  distinct_countries  distinct_languages
0        107.947931                 7  ...                  65                  48

[1 rows x 11 columns]

In our initial exploration of the movie dataset, we saw 4968 total records across several key columns, including film titles, release years, countries, durations, languages, certifications, gross revenues, and budgets. We identified missing values in several columns, notably 'release_year', 'duration', 'gross', and 'budget'. Our dataset variables include a mix of data types: integers, character-varying (strings), and bigints. Numerical variables displayed a wide range of values, such as 'duration' spanning from 7 to 334 minutes and 'gross' spanning from \$162 to \$936,627,416. This preliminary analysis provides a solid starting point for our subsequent analysis and is vital for guiding us in understanding the factors that contribute to the success and characteristics of these films.

## 2.4. Insights and Data Visualization

### 1. How many movies are present in the database?

In [24]:
-- Calculate the total number of movies in the films table
SELECT COUNT(*) AS total_movies
FROM cinema.films;

total_movies
0          4968

The database contains a total of **4,968** movies. This figure represents the number of films cataloged in the `cinema.films` table, providing a substantial dataset for analysis. The diverse collection of movies included could be leveraged to identify trends and patterns in film production and success, aiding the studio in making informed decisions for future projects aimed at recapturing the essence of classic Hollywood cinema.

### 2. There seems to be a lot of missing data in the gross and budget columns. How many rows have missing data? What would you recommend your manager do with these rows?

In [25]:
-- Calculate total rows, count of non-missing and missing values for gross and budget columns in the films table
SELECT 
	COUNT(*) AS total_rows,
	COUNT(gross) AS non_missing_gross,
	COUNT(budget) AS non_missing_budget,
	COUNT(*) - COUNT(gross) AS missing_gross,
	COUNT(*) - COUNT(budget) AS missing_budget
FROM cinema.films;

total_rows  non_missing_gross  ...  missing_gross  missing_budget
0        4968               4158  ...            810             430

[1 rows x 5 columns]

The analysis of the `cinema.films` database reveals that **out of a total of 4,968** movies, **810 entries lack data in the `gross` revenue column**, and **430 entries are missing `budget` information**. This suggests a significant number of films, approximately 16.3% for `gross` and 8.7% for `budget`, do not have complete financial data. Such gaps in key financial metrics could potentially skew any analysis related to profitability or financial success. Addressing these gaps is crucial for a comprehensive analysis of financial performance. To mitigate the impact of these missing values, it might be recommended to either exclude these films from specific financial analyses or use statistical methods such as imputation to estimate the missing data based on available information, thus allowing for more complete and robust analysis.

### 3. How many different certifications or ratings are present in the database?

In [26]:
-- Retrieve distinct movie certifications and order them alphabetically
SELECT DISTINCT certification
FROM cinema.films
ORDER BY certification;

certification
0               
1       Approved
2              G
3             GP
4              M
5          NC-17
6      Not Rated
7         Passed
8             PG
9          PG-13
10             R
11       Unrated
12             X

The `cinema.films` database includes a variety of film certifications, reflecting different audience suitability ratings, with a total of 13 distinct types identified. These certifications span from early classifications such as "Approved," "Passed," and "M" (Mature), to the modern ratings used by the [Motion Picture Association (MPA)](https://simple.wikipedia.org/wiki/Motion_Picture_Association_film_rating_system): **G (General Audiences)**, **PG (Parental Guidance Suggested)**, **PG-13 (Parents Strongly Cautioned)**, **R (Restricted)**, and **NC-17 (Adults Only)**. Additionally, the dataset encompasses transitional ratings like "GP" (later changed to PG) and "X" (predating NC-17), as well as "Not Rated," "Unrated," and a missing rating (' '), indicating films released without an official MPA rating. The evolution of these ratings reflects changing societal values and regulatory standards, transitioning from the [Hays Code's](https://en.wikipedia.org/wiki/Hays_Code) "Approved" and "Passed" ratings to the detailed and age-specific classifications of the modern system introduced in 1968. The addition of "PG-13" in 1984 bridged the gap between "PG" and "R," while the replacement of "X" with "NC-17" in 1990 distinguished legitimate adult films from pornography. This diversity in film ratings underscores the broad range of content and intended audiences in the historical film data, providing insights into how films have been classified and marketed over the years.

### 4. What are the top five countries in terms of number of movies produced?

In [27]:
-- Retrieve the top 5 countries by the number of movies produced, ordered by the count in descending order
SELECT
	country, 
	COUNT(*) AS number_of_movies
FROM cinema.films
GROUP BY country
ORDER BY number_of_movies DESC
LIMIT 5;

country  number_of_movies
0      USA              3750
1       UK               443
2   France               153
3   Canada               123
4  Germany                97

The analysis of the `cinema.films` database reveals that the **United States leads** substantially in film production **with a total of 3,750 movies**, underscoring its dominant position in the global cinema industry. The **United Kingdom** follows as a distant second, having **produced 443 films**. **France**, **Canada**, and **Germany** also make significant contributions to the film industry, **with 153**, **123**, and **97 movies produced respectively**. This data highlights the major hubs of movie production prior to 1997, with these top five countries playing pivotal roles in the cinematic landscape, influencing both cultural narratives and industry trends internationally.

### 5. What is the average duration of English versus French movies?

In [28]:
-- Calculate the average duration of movies in English and French, grouped by language
SELECT
	language,
  	AVG(duration) AS average_duration_in_minutes
FROM cinema.films
WHERE language IN ('English', 'French')
GROUP BY language;

language  average_duration_in_minutes
0   French                   104.666667
1  English                   107.636541

The result reveals that **English-language movies** have an average duration of approximately **107.64 minutes**, slightly surpassing **French films**, which average around **104.67 minutes**. This results in **a difference of about 2.97 minutes**, with English movies generally having longer runtimes. This subtle distinction suggests that English films may incorporate additional content or narrative elements, reflecting different storytelling styles or production practices between the two cultures. These insights are critical for understanding cultural nuances in filmmaking, which could inform strategic decisions in future film productions aimed at capturing the essence of both cinematic traditions.

### 6. Other questions I explored during my analysis

In addition to the main questions, here are some additional questions that I explored using the data:

1. How many movies were produced per decade?
2. What were the average budgets per decade?
3. What was the average gross revenue of movies per decade?
4. What percentage of movies were profitable?
5. Which films released before 1977 have the highest return on investment (ROI) in our database?
6. What are the top ten movies with the highest gross in our database that were released before 1997?

#### 1. How many movies were produced per decade?

In [29]:
-- Categorize movies by decade and count the number of movies per decade
WITH year_range_data AS (
    SELECT
        CASE
            WHEN release_year = 1916 THEN '1916'
            WHEN release_year BETWEEN 1920 AND 1929 THEN '1920-1929'
            WHEN release_year BETWEEN 1930 AND 1939 THEN '1930-1939'
            WHEN release_year BETWEEN 1940 AND 1949 THEN '1940-1949'
            WHEN release_year BETWEEN 1950 AND 1959 THEN '1950-1959'
            WHEN release_year BETWEEN 1960 AND 1969 THEN '1960-1969'
            WHEN release_year BETWEEN 1970 AND 1979 THEN '1970-1979'
            WHEN release_year BETWEEN 1980 AND 1989 THEN '1980-1989'
            WHEN release_year BETWEEN 1990 AND 1999 THEN '1990-1999'
            WHEN release_year BETWEEN 2000 AND 2009 THEN '2000-2009'
            WHEN release_year BETWEEN 2010 AND 2016 THEN '2010-2016'
        END AS year_range
    FROM
        cinema.films
    WHERE
        release_year IS NOT NULL
)
SELECT
    year_range,
    COUNT(*) AS number_of_movies
FROM
    year_range_data
GROUP BY
    year_range
ORDER BY
    year_range;

year_range  number_of_movies
0        1916                 1
1   1920-1929                 5
2   1930-1939                15
3   1940-1949                25
4   1950-1959                28
5   1960-1969                72
6   1970-1979               114
7   1980-1989               291
8   1990-1999               786
9   2000-2009              2098
10  2010-2016              1491

The earliest recorded year of release, **1916, features only one film**, while the subsequent **1920s hold five movies**. The **1930s**, a period marked by the growth of talkies and the Golden Age of Hollywood, saw an **increase to 15 films**. This trend continued steadily through the **1940s and 1950s**, which **had 25 and 28 movies**, respectively.

Cinematic production began to expand more significantly in the **1960s**, which **contained 72 movies**. The **1970s**, driven by creative experimentation, **nearly doubled this output to 114 films**. By the **1980s**, a substantial **increase to 291 films** reflected the rise of blockbuster cinema. The **1990s**, a golden era for diverse genres, **produced 786 movies**. **Post-1997**, the output of films accelerates dramatically, with the **2000s containing 2,098 movies** and the **2010s holding 1,491 movies**.

#### 2. What were the average budgets per decade?

In [30]:
-- Categorize movies by decade, and calculate the average budget for each decade
SELECT
    CASE
        WHEN release_year = 1916 THEN '1916'
        WHEN release_year BETWEEN 1920 AND 1929 THEN '1920-1929'
        WHEN release_year BETWEEN 1930 AND 1939 THEN '1930-1939'
        WHEN release_year BETWEEN 1940 AND 1949 THEN '1940-1949'
        WHEN release_year BETWEEN 1950 AND 1959 THEN '1950-1959'
        WHEN release_year BETWEEN 1960 AND 1969 THEN '1960-1969'
        WHEN release_year BETWEEN 1970 AND 1979 THEN '1970-1979'
        WHEN release_year BETWEEN 1980 AND 1989 THEN '1980-1989'
        WHEN release_year BETWEEN 1990 AND 1999 THEN '1990-1999'
        WHEN release_year BETWEEN 2000 AND 2009 THEN '2000-2009'
        WHEN release_year BETWEEN 2010 AND 2016 THEN '2010-2016'
    END AS year_range,
    AVG(budget) AS average_budget_in_usd
FROM
    cinema.films
WHERE
    release_year BETWEEN 1916 AND 2016
    AND budget IS NOT NULL
    AND budget > 0
GROUP BY
    year_range
ORDER BY
    year_range;

year_range  average_budget_in_usd
0        1916           3.859070e+05
1   1920-1929           1.681000e+06
2   1930-1939           1.638910e+06
3   1940-1949           2.388827e+06
4   1950-1959           2.970456e+06
5   1960-1969           5.492874e+06
6   1970-1979           7.913097e+06
7   1980-1989           1.822726e+07
8   1990-1999           3.741215e+07
9   2000-2009           4.765436e+07
10  2010-2016           4.107224e+07

The solitary film from **1916** had a modest budget of \$385,907. By the **1920s**, as Hollywood emerged as a global cinema hub, budgets rose to an average of \$1.68 million, and the introduction of sound in the **1930s** maintained this level with a slight dip to \$1.64 million.

The **1940s** saw an upward trend with an average budget of \$2.39 million, indicating increased studio investment despite World War II. In the **1950s**, budgets climbed to \$2.97 million, partly driven by widescreen innovations and the rise of color films. The **1960s** saw significant jumps to \$5.49 million, coinciding with the studio system's decline and the emergence of independent filmmakers.

Budgets averaged \$7.91 million in the **1970s**, as blockbusters like Star Wars and Jaws dominated. The **1980s** saw a doubling of investment to \$18.23 million, reflecting Hollywood's appetite for high-concept franchises. By the **1990s**, budgets had soared to \$37.41 million, ushering in a new era of CGI and expansive storytelling. The exponential growth continued into the **2000s**, with budgets reaching \$47.65 million, before stabilizing in the **2010s** at an average of \$41.07 million, possibly due to a shift towards digital distribution and increased emphasis on profitability.

#### 3. What was the average gross revenue of movies per decade?

In [31]:
-- Categorize movies by decade and calculate the average gross for each decade
WITH year_range_data AS (
    SELECT
        CASE
			WHEN release_year = 1916 THEN '1916'
			WHEN release_year BETWEEN 1920 AND 1929 THEN '1920-1929'
			WHEN release_year BETWEEN 1930 AND 1939 THEN '1930-1939'
			WHEN release_year BETWEEN 1940 AND 1949 THEN '1940-1949'
			WHEN release_year BETWEEN 1950 AND 1959 THEN '1950-1959'
			WHEN release_year BETWEEN 1960 AND 1969 THEN '1960-1969'
			WHEN release_year BETWEEN 1970 AND 1979 THEN '1970-1979'
			WHEN release_year BETWEEN 1980 AND 1989 THEN '1980-1989'
			WHEN release_year BETWEEN 1990 AND 1999 THEN '1990-1999'
			WHEN release_year BETWEEN 2000 AND 2009 THEN '2000-2009'
			WHEN release_year BETWEEN 2010 AND 2016 THEN '2010-2016'
        END AS year_range,
        gross
    FROM
        cinema.films
    WHERE
        release_year IS NOT NULL
        AND gross IS NOT NULL
)
SELECT
    year_range,
    AVG(gross) AS average_gross_in_usd
FROM
    year_range_data
GROUP BY
    year_range
ORDER BY
    year_range;

year_range  average_gross_in_usd
0  1920-1929          1.461096e+06
1  1930-1939          6.854110e+07
2  1940-1949          4.435873e+07
3  1950-1959          1.838363e+07
4  1960-1969          4.201587e+07
5  1970-1979          7.070075e+07
6  1980-1989          5.170222e+07
7  1990-1999          4.538012e+07
8  2000-2009          4.345259e+07
9  2010-2016          5.872645e+07

In the **1920s**, the earliest era in our dataset, movie gross earnings averaged \$1.46 million, reflecting the fledgling industry's modest financial scale. However, the **1930s** saw a dramatic surge to an average of \$68.54 million per film, driven by Hollywood's Golden Age and the booming popularity of the silver screen.

The **1940s** saw a dip to \$44.36 million, influenced by World War II's constraints, while the post-war **1950s** reflected a further decline to \$18.38 million as television began to reshape entertainment consumption. Gross earnings rebounded in the **1960s**, averaging \$42.02 million per film, fueled by the rise of international markets and artistic experimentation.

The **1970s** continued this upward trend, reaching \$70.70 million, largely due to the emergence of the blockbuster era with Jaws and Star Wars. In the **1980s**, a cultural shift towards high-concept action movies and franchise filmmaking brought in an average of \$51.70 million per film. The **1990s** maintained substantial earnings with \$45.38 million, while the **2000s** averaged \$43.45 million per film. The **2010s** capped the analysis at an average of \$58.73 million per film, signifying the era's continued investment in high-grossing franchises and global market expansion.

#### 4. What percentage of movies were profitable?

In [32]:
-- Calculate the number and percentage of profitable movies
WITH profit_data AS (
    SELECT
        title, 
        (gross - budget) AS profit, 
        gross,
        budget
    FROM cinema.films
    WHERE gross IS NOT NULL AND budget IS NOT NULL AND budget > 0
)
SELECT 
    COUNT(CASE WHEN profit > 0 THEN 1 ELSE NULL END) AS profitable_movies,
    COUNT(*) AS total_movies,
    ROUND((COUNT(CASE WHEN profit > 0 THEN 1 ELSE NULL END)::numeric / COUNT(*)::numeric) * 100, 2) AS percentage_profitable
FROM profit_data;

profitable_movies  total_movies  percentage_profitable
0               2051          3892                   52.7

The analysis of profitability in the `cinema.films` dataset reveals that **out of 3,892 movies with both gross and budget data** available, **2,051 films, or 52.7%, were profitable**. This majority demonstrates the potential financial viability of filmmaking, with more than half of the movies recouping their budgets and earning additional revenue. The **remaining 47.3% of movies that were unprofitable** highlight the inherent risks of the film industry, where high production costs, market competition, and shifting audience preferences can significantly impact commercial success.

#### 5. What are the top ten movies with the highest gross in our database that were released before 1997?

In [33]:
-- Retrieve the top 10 highest-grossing movies released before 1997, including title, release year, certification, language, country, and gross
SELECT
	title,
	release_year,
	certification,
	language, 
	country,
	gross AS gross_usd
FROM cinema.films
WHERE release_year < 1997 AND gross IS NOT NULL
ORDER BY gross DESC
LIMIT 10;

title  ...  gross_usd
0              Star Wars: Episode IV - A New Hope  ...  460935665
1                      E.T. the Extra-Terrestrial  ...  434949459
2                                   The Lion King  ...  422783777
3                                   Jurassic Park  ...  356784000
4                                    Forrest Gump  ...  329691196
5      Star Wars: Episode VI - Return of the Jedi  ...  309125409
6                                Independence Day  ...  306124059
7  Star Wars: Episode V - The Empire Strikes Back  ...  290158751
8                                      Home Alone  ...  285761243
9                                            Jaws  ...  260000000

[10 rows x 6 columns]

![star_wars_ver2_xlg](star_wars_ver2_xlg.jpg "Movie poster of Star Wars: Episode IV - A New Hope (1977)")

_Movie poster of "Star Wars: Episode IV - A New Hope (1977)." Image from_ [IMP Awards](http://www.impawards.com/1977/star_wars_ver2.html)

The top ten highest-grossing movies in the database that were released before 1997 showcase a collection of iconic films spanning two decades, from 1975 to 1996. Leading the list is **Star Wars: Episode IV - A New Hope (1977)**, which revolutionized the sci-fi genre, grossing an impressive \$460.9 million. **E.T. the Extra-Terrestrial (1982)** follows with \$434.9 million, capturing hearts worldwide. Disney's animated triumph, **The Lion King (1994)**, stands out with \$422.8 million, demonstrating the enduring appeal of family-friendly adventures.

**Jurassic Park (1993**) brought dinosaurs back to life with \$356.8 million, while **Forrest Gump (1994)** enthralled audiences with a touching story, earning \$329.7 million. The space saga, **Star Wars: Episode VI - Return of the Jedi (1983)**, and the explosive action-packed, **Independence Day (1996)** secured their positions with \$309.1 million and \$306.1 million, respectively.

Further securing the dominance of the Star Wars franchise, **Star Wars: Episode V - The Empire Strikes Back (1980)** grossed \$290.2 million. **Home Alone (1990)** showcased the comedy's profitability at \$285.8 million, and **Jaws (1975)**, the shark-infested blockbuster thriller, remains remarkable at \$260 million.

All ten films are primarily English-language productions from the United States, highlighting Hollywood's significant influence on global cinema during this era.

#### 6. Which films released before 1977 have the highest return on investment (ROI) in our database?

To find out which movie was profitable, we compare their return on investment (ROI). The ROI is calculated using the formula:

$$ \text{ROI} = \frac{\text{gross} - \text{budget}}{\text{budget}} \times 100 $$

In [34]:
-- Retrieve the top 10 movies released before 1977 with the highest return on investment (ROI), including title, release year, certification, language, country, gross, budget, and ROI
SELECT 
	DISTINCT title,
    release_year, 
	certification,
	language,
	country,
    gross, 
    budget, 
    ((gross - budget) / budget) * 100 AS roi
FROM cinema.films
WHERE release_year < 1977 
    AND gross IS NOT NULL 
    AND budget IS NOT NULL 
    AND budget > 0
ORDER BY roi DESC
LIMIT 10;

title  release_year  ...   budget    roi
0     The Texas Chain Saw Massacre          1974  ...    83532  36800
1                American Graffiti          1973  ...   777000  14700
2                            Rocky          1976  ...   960000  12100
3  Snow White and the Seven Dwarfs          1937  ...  2000000   9100
4                            Benji          1974  ...   500000   7800
5               Gone with the Wind          1939  ...  3977000   4800
6                  Blazing Saddles          1974  ...  2600000   4400
7                           Psycho          1960  ...   806947   3800
8                         Fantasia          1940  ...  2280000   3200
9                             Jaws          1975  ...  8000000   3100

[10 rows x 8 columns]

![texas_chainsaw_massacre_xlg](texas_chainsaw_massacre_xlg.jpg "Movie poster of The Texas Chainsaw Massacre (1974)")

_Movie poster of "The Texas Chainsaw Massacre (1974)." Image from_ [IMP Awards](http://www.impawards.com/1974/texas_chainsaw_massacre.html)

The top ten films released before 1977 with the highest return on investment (ROI) illustrate a diverse collection of creative works that have secured their legacy by achieving remarkable profitability. Leading this illustrious group is **The Texas Chain Saw Massacre (1974)**, which astounded audiences with its gritty horror on a minimal budget, achieving a staggering ROI of 36,800%, turning a mere \$83,532 budget into \$30.86 million. **American Graffiti (1973)** follows closely with a 14,700% ROI, thanks to George Lucas' nostalgic and relatable depiction of teenage life, grossing \$115 million on a budget of only \$777,000.

**Rocky (1976)**, the underdog sports drama starring Sylvester Stallone, earned an impressive 12,100% ROI, proving that determination and grit could resonate widely, converting a \$960,000 budget into \$117.24 million. Walt Disney's **Snow White and the Seven Dwarfs (1937)**, the first feature-length animated film, remains a testament to timeless storytelling, securing a 9,100% ROI with its \$2 million budget yielding \$184.93 million.

**Benji (1974)**, a family-friendly tale of a lovable stray dog, won hearts worldwide with its \$39.55 million gross and 7,800% ROI. The epic classic **Gone with the Wind (1939)** secured a 4,800% ROI, turning a \$3.98 million budget into \$198.66 million. Mel Brooks' satirical comedy **Blazing Saddles (1974)** lampooned Westerns while achieving a 4,400% ROI, with a \$2.6 million budget producing \$119.5 million.

Alfred Hitchcock's **Psycho (1960)** disrupted conventional storytelling and terrified audiences, garnering a 3,800% ROI with \$32 million from an \$806,947 budget. Disney's orchestral marvel **Fantasia (1940)** blended animation and classical music for a mesmerizing cinematic experience, securing a 3,200% ROI. Finally, **Jaws (1975)** achieved an impressive 3,100% ROI, grossing \$260 million from an \$8 million budget.

## 2.5. Conclusion

The analysis of the `cinema.films` dataset has provided vital insights that can significantly inspire future film production strategies aimed at recapturing the magic of old Hollywood. With a dataset spanning several decades, key findings such as the total number of films, the variety of certifications, and the top-producing countries provide a deep understanding of historical film trends. Notably, the United States emerged as the predominant film-producing nation with a vast majority of cinematic contributions, followed by the United Kingdom, France, Canada, and Germany. This global diversity exemplifies the broad appeal and varied tastes of cinemagoers. Moreover, the detailed exploration of financial metrics revealed significant gaps in gross and budget data, prompting recommendations for sophisticated data imputation methods to ensure analyses remain robust and comprehensive.

Additionally, the exploration of average movie durations across different languages, profitability rates, and top-grossing films offers actionable insights for studio executives. For instance, the slightly longer duration of English films compared to French suggests a potential for more elaborate or expansive storytelling in English-speaking markets. Furthermore, with over half of the movies proving profitable, there's a strong case for continuing traditional cinematic themes and narratives that have historically resonated with audiences. However, the high ROI figures for certain classics highlight the potential for blockbuster returns on modest budgets, particularly in genres that resonate deeply with viewers, such as horror and drama. The profitability of iconic films like The Texas Chain Saw Massacre (1974), Rocky (1976), and American Graffiti (1973) proves that compelling storytelling, even on smaller budgets, can yield remarkable financial success. This trend demonstrates the enduring appeal of innovative, creative, and high-impact filmmaking, which modern studios can emulate by investing in stories that resonate with audiences emotionally and culturally.

The blockbuster hits of the 1970s through the 1990s reaffirmed the ability of high-concept franchises to draw vast international audiences, while the overall increase in movie budgets and gross earnings across the decades underscored a period of significant technological advancement and creative ambition. Notably, the 1990s, with the introduction of CGI and expansive storytelling, became a landmark period for cinematic diversity and innovation. These findings reflect the shifting dynamics of film production, where creative risk-taking and technical prowess must align with market expectations. The comprehensive examination of historical data thus informs future productions to blend timeless storytelling techniques with modern technology and market demands, ensuring that upcoming films resonate across generations and create new cinematic legacies.

## 2.6. Recommendations

Based on the insights garnered from the analysis of the movie dataset, the following recommendations are proposed to inspire future film productions:

1.  **Target Proven Storytelling Formats:** 
      - Leverage the appeal of genres like horror, drama, and adventure, which historically have shown high returns on investment. 
      - Draw inspiration from successful blockbusters, emphasizing imaginative storytelling, memorable characters, and universal themes.
2. **Balanced Budget with Expected ROI:**
      - Avoid overextending budgets by focusing on creative and strategic storytelling, similar to classics like The Texas Chain Saw Massacre (1974) and American Graffiti (1973), which achieved remarkable returns.
      - Prioritize quality and originality over sheer scale, as smaller-budget films often yield significant returns.
3. **Blend Technology with Traditional Themes:**
      - Incorporate CGI, immersive sound, and other modern technological advancements to enhance storytelling, while still honoring traditional cinematic techniques.
      - Revisit themes and motifs that have proven timeless, such as underdog stories like Rocky (1976) and epic adventures like the Star Wars franchise, while integrating them with contemporary technological storytelling methods.
4. **Adapt and Diversify Certification Strategies:**
      - Tailor films for specific audience demographics by considering diverse certification strategies to maximize market reach.
      - Explore creating content across different certification levels to widen audience appeal, from family-friendly The Lion King (1996) to adult Psycho (1960).
5. **Continuous Trend Monitoring:**
      - Stay informed about evolving audience preferences and technological advancements by regularly analyzing market trends and comparing them to historical data.
      - Encourage innovation by supporting diverse filmmaking voices and fostering creative experimentation.

## 2.7. Links/References

1. Data Source: IMDb: Ratings, Reviews, and Where to Watch the Best Movies & TV Shows. (n.d.). IMDb. https://www.imdb.com/
2. Image Cover: Jones, J. (2023, April 15). Jaws: The Creation Of Hollywood’s Most Iconic Movie Poster. All the Right Movies. https://alltherightmovies.com/feature/jaws-the-creation-of-hollywoods-most-iconic-movie-poster/
3. Motion Picture Association film rating system. (2024, March 8). Simple English Wikipedia, the Free Encyclopedia. https://simple.wikipedia.org/wiki/Motion_Picture_Association_film_rating_system
4. Hays Code. (2024, April 7). Wikipedia. https://en.wikipedia.org/wiki/Hays_Code
5. Star Wars Movie Poster (#2 of 16). (n.d.). IMP Awards. http://www.impawards.com/1977/star_wars_ver2.html
6. The Texas Chainsaw Massacre Movie Poster (#1 of 4). (n.d.). IMP Awards. http://www.impawards.com/1974/texas_chainsaw_massacre.html

## Resources

Tools Used:

1. ChatGPT: Code Generator, Brainstormer
2. Quillbot: Grammar Checker, Paraphraser, Citation Generator